django/forms/fields.py

```diff
@@ -723,7 +723,7 @@ def has_changed(self, initial, data):
         if initial == 'False':
             # show_hidden_initial may have transformed False to 'False'
             initial = False
-        return bool(initial) != bool(data)
+        return bool(initial) != self.to_python(data)
```

Comment:
@claudep, totally agree, when looking at the code there it seemed like it would help prevent future issues to liberally use `to_python`. I opted to keep the changes small and concise so they were more likely to be accepted, but I'd be a fan of also running initial through `to_python` here. If we do that on `BooleanField` you could really just drop the implementation of `has_changed` on `NullBooleanField`.

Harmonizing `to_python` between them also seems like a good idea, but might be a different issue to address. There would be a behavior change there since currently `BooleanField` considers '2' to be `True` but `NullBooleanField` considers it to be `None`, so I think that may take a bit more thought on how to harmonize it and not break current behavior.